### PR TITLE
Convert alerts to use event streams, update tests

### LINKED ISSUE #149: Alerts should use event streams
Currently the alerts component uses "arbitrary" timeouts during `hideAlert` in order to wait for CSS animations to finish before removing it from the list. This could be improved.

### LINKED ISSUE #154: Replace "hub" module with can-event
The hub module is currently only being used by the error-handler behavior.

### PR DESCRIPTION
**Pull request by DesignByOnyx**
Closes #149 
Closes #154

### CODE REVIEW

**Comment by DesignByOnyx** [comment]
Please merge #212 before merging this into staging.

File: public/components/alerts/alerts.js [hunk 1]
```diff
@@ -1,57 +1,94 @@
...
+            return alerts;
+        }
+      }, []);
     }
   },
-  hideAlert (alert) {
-    alert.visible = false;
-    // TODO: this should happen on CSS animation end
-    setTimeout(() => this.removeAlert(alert), 500);
+  addAlert (alert) {
```
Comment (justinbmeyer) [inline public/components/alerts/alerts.js]: remove this.

File: public/components/alerts/alerts.js [hunk 2]
```diff
@@ -1,57 +1,94 @@
...
+  }
+};
 
 export const ViewModel = DefineMap.extend({
-  subscription: '*',
-  alerts: {
-    Value: AlertItem.List
+  addAlertStream: {
+    value() {
+      return this.stream('add-alert')
```
Comment (justinbmeyer) [inline public/components/alerts/alerts.js]: no need to listen to add-alert

File: public/components/alerts/alerts.js [hunk 3]
```diff
@@ -1,57 +1,94 @@
...
+  remove(alerts, alert) {
+    return alerts.filter(item => item.id !== alert.id);
+  }
+};
 
 export const ViewModel = DefineMap.extend({
-  subscription: '*',
-  alerts: {
-    Value: AlertItem.List
+  addAlertStream: {
```
Comment (justinbmeyer) [inline public/components/alerts/alerts.js]: can probably be removed

File: public/components/alerts/alerts.js [hunk 4]
```diff
@@ -1,57 +1,94 @@
...
+        } 
+        return Kefir.constant({ action: 'no-op' });
+      });
+    }
   },
-  removeAlert (alert) {
-    const idx = this.alerts.indexOf(alert);
-    if (idx !== -1) {
-      this.alerts.splice(idx, 1);
+  removeAlertStream: {
```
Comment (justinbmeyer) [inline public/components/alerts/alerts.js]: can probably be removed

File: public/components/alerts/alerts.js [hunk 5]
```diff
@@ -1,57 +1,94 @@
...
+            alert
+          };
+        });
+    }
   },
-  addAlert (alert) {
-    this.alerts.push(alert);
+  autoHideStream: {
+    value() {
+      return this.addAlertStream.flatMap(ev => {
```
Comment (justinbmeyer) [inline public/components/alerts/alerts.js]: `this.addAlertStream` -> `this.stream('alert')`

File: public/components/alerts/alerts.js [hunk 6]
```diff
@@ -1,57 +1,94 @@
...
+        if (ev.alert.displayInterval > 0) {
+          return Kefir.later(ev.alert.displayInterval, {
+            action: 'REMOVE_ALERT',
```
Comment (justinbmeyer) [inline public/components/alerts/alerts.js]: `type: 'remove'`

File: public/components/alerts/alerts.js [hunk 7]
```diff
@@ -1,57 +1,94 @@
...
+            alert: ev.alert
+          });
+        } 
+        return Kefir.constant({ action: 'no-op' });
```
Comment (justinbmeyer) [inline public/components/alerts/alerts.js]: type: 'no-op'

File: public/components/alerts/alerts.js [hunk 8]
```diff
@@ -1,57 +1,94 @@
...
-    setTimeout(() => {
-      alert.visible = true;
-    }, 30);
-    
-    if (alert.displayInterval > 0) {
-      setTimeout(() => this.hideAlert(alert), 30 + alert.displayInterval);
+  alerts: {
+    stream() {
+      // TODO: use can-event
+      return Kefir.merge([this.addAlertStream, this.autoHideStream, this.removeAlertStream]).scan((alerts, ev) => {
```
Comment (justinbmeyer) [inline public/components/alerts/alerts.js]: `this.stream("alert").merge( this. autoHideStream ).merge( this.stream("remove") )`

File: public/components/alerts/alerts.js [hunk 9]
```diff
@@ -1,57 +1,94 @@
...
+        let alert = ev.alert;
+        switch (ev.action) {
```
Comment (justinbmeyer) [inline public/components/alerts/alerts.js]: `ev.type`

File: public/components/alerts/alerts.js [hunk 10]
```diff
@@ -1,57 +1,94 @@
...
+import view from './alerts.stache';
 import './alerts.less';
 import hub from '../../lib/hub';
 import AlertItem from '../../models/alert';
 
-/**
- * TODO: use streams for modifying alerts
- *      - https://github.com/donejs/bitcentive/issues/149
- */
+export const reducers = {
```
Comment (justinbmeyer) [inline public/components/alerts/alerts.js]: move this into `.scan`

File: public/components/alerts/alerts.js [hunk 11]
```diff
@@ -1,57 +1,61 @@
...
-import AlertItem from '../../models/alert';
+import hub from '../../lib/event-hub';
+import CID from 'can-cid';
 
-/**
- * TODO: use streams for modifying alerts
- *      - https://github.com/donejs/bitcentive/issues/149
- */
+
+var alertStream = CanStream.toStream(hub, 'alert').map(ev => {
```
Comment (justinbmeyer) [inline public/components/alerts/alerts.js]: `canStream`

File: public/components/alerts/alerts.js [hunk 12]
```diff
@@ -1,57 +1,61 @@
...
+  autoHideStream: {
+    value() {
+      return alertStream.flatMap(alert => {
+        if (alert.displayInterval > 0) {
+          return Kefir.later(alert.displayInterval, {
+            type: 'remove',
+            id: alert.id
+          });
+        } 
+        return Kefir.constant({ type: 'no-op' });
```
Comment (justinbmeyer) [inline public/components/alerts/alerts.js]: is there a way to make it so there is no event?

File: public/components/alerts/alerts_test.js [hunk 13]
```diff
@@ -1,68 +1,54 @@
...
+	const handler = ev => {
+		assert.equal(ev.type, 'remove');
+		assert.equal(ev.id, 1234);
+		vm.off('remove', handler);
 		done();
-	}, 400);
+	};
+
+	vm.on('remove', handler);
+	vm.removeAlert({ id: 1234 });
```
Comment (justinbmeyer) [inline public/components/alerts/alerts_test.js]: make sure the alert is removed from `.alerts`

File: public/components/alerts/alerts.js (hunk 12 above)
Comment (DesignByOnyx) [inline public/components/alerts/alerts.js]: I couldn't figure out a way around this as flatMap has to return _something_. I can filter out the 'no-op' events if you want - akin to `[].map(...).filter(...)`.

File: public/components/alerts/alerts.md [hunk 14]
```diff
@@ -1,15 +1,17 @@
...
-hub.publish('alert', {
-	type: 'info',
-	title: 'Something happened!',
-	message: 'Please run npm install'
+hub.dispatch({
+	type: 'alert',
+	kind: 'danger|success|info|warning',
+	title: 'Optional title',
+	message: 'Please run npm install',
+	displayInterval: -1 // ms - used to autohide alert
```
Comment (justinbmeyer) [inline public/components/alerts/alerts.md]: Infinity?

**Pull request merged**